<a href="https://colab.research.google.com/github/joeyMckinney/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/Josiah_McKinney_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [62]:
### YOUR WORK HERE
import numpy as np
import pandas as pd
from scipy import stats

column_headers = ['symboling', 'normalized_losses', 'make', 'fuel-type', 
                  'aspiration', 'num_of_doors', 'body_style', 'drive_wheels', 
                  'engine_location', 'wheel_base', 'length', 'width', 'height', 
                  'curb-weight', 'engine_type', 'num_of_cylinders', 
                  'engine_size', 'fuel_system', 'bore', 'stroke', 
                  'compression_ratio', 'horsepower', 'peak_rpm', 'city_mpg', 
                  'highway_mpg', 'price']
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data', names=column_headers, na_values='?')


In [63]:
df.head()

symboling  normalized_losses         make  ... city_mpg highway_mpg    price
0          3                NaN  alfa-romero  ...       21          27  13495.0
1          3                NaN  alfa-romero  ...       21          27  16500.0
2          1                NaN  alfa-romero  ...       19          26  16500.0
3          2              164.0         audi  ...       24          30  13950.0
4          2              164.0         audi  ...       18          22  17450.0

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [64]:
### YOUR WORK HERE
sample_cars = df.sample(10, random_state=30)
sample_cars.head(10)

symboling  normalized_losses  ... highway_mpg    price
121          1              154.0  ...          38   6692.0
176         -1               65.0  ...          32  10898.0
24           1              148.0  ...          38   6229.0
88          -1              137.0  ...          30   9279.0
72           3              142.0  ...          18  35056.0
103          0              108.0  ...          25  13499.0
26           1              148.0  ...          38   7609.0
20           0               81.0  ...          43   6575.0
155          0               91.0  ...          32   8778.0
185          2               94.0  ...          34   8195.0

[10 rows x 26 columns]

In [65]:
print(df['highway_mpg'].mean())
print(smaple_cars['highway_mpg'].mean())
print('yes they are different')

30.75121951219512
32.8
yes they are different


In [69]:
stats.ttest_1samp(sample_cars['highway_mpg'], df['highway_mpg'].mean())

Ttest_1sampResult(statistic=0.8912422391619923, pvalue=0.39600040771093326)

based on a t-statistic of .8912, and a p-vlaue of .396 > .05, i fail reject the null hypothesis that the sample mean and the population mean are the same.






---



The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [77]:
### YOUR WORK HERE
stats.ttest_1samp(sample_cars['highway_mpg'], 35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

based on a t-statistic of .8912, and a p-vlaue of .396 > .05, i fail to reject the null hypothesis that the sample mean and the population mean are the same.

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [79]:
### YOUR WORK HERE
sample_cars_100 = df.sample(100, random_state=30)
sample_cars_100.shape

(100, 26)

In [80]:
stats.ttest_1samp(sample_cars_100['highway_mpg'], 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

based on a t-statistic of -0.166, and a p-vlaue of approx .0, i reject the null hypothesis that the sample mean and the population mean are the same.

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

because the smaller the sample size the more liky change that you are wrong. and by wrong i have a less change of representing the whole population.

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [83]:
### YOUR WORK HERE
cols = [
        'Party',
        'handicapped-infants',
        'water-project-cost-sharing',
        'adoption-of-the-budget-resolution',
        'physician-fee-freeze',
        'el-salvador-aid',
        'religious-groups-in-schools',
        'anti-satellite-test-ban',
        'aid-to-nicaraguan-contras',
        'mx-missile',
        'immigration',
        'synfuels-corporation-cutback',
        'education-spending',
        'superfund-right-to-sue',
        'crime',
        'duty-free-exports',
        'export-administration-act-south-africa',
        ]
df2 = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data', names=cols, na_values='?')
df2.head()


Party  ... export-administration-act-south-africa
0  republican  ...                                      y
1  republican  ...                                    NaN
2    democrat  ...                                      n
3    democrat  ...                                      y
4    democrat  ...                                      y

[5 rows x 17 columns]

In [86]:
df2 = df2.replace({'y' : 1, 'n' : 0})
df2.head()

Party  ...  export-administration-act-south-africa
0  republican  ...                                     1.0
1  republican  ...                                     NaN
2    democrat  ...                                     0.0
3    democrat  ...                                     1.0
4    democrat  ...                                     1.0

[5 rows x 17 columns]

In [88]:
rep = df2[df2['Party']=='republican']
dem = df2[df2['Party']=='democrat']
print(rep.shape)
rep.head()

(168, 17)


Party  ...  export-administration-act-south-africa
0   republican  ...                                     1.0
1   republican  ...                                     NaN
7   republican  ...                                     1.0
8   republican  ...                                     1.0
10  republican  ...                                     0.0

[5 rows x 17 columns]

In [89]:
print(dem.shape)
dem.head()

(267, 17)


Party  ...  export-administration-act-south-africa
2  democrat  ...                                     0.0
3  democrat  ...                                     1.0
4  democrat  ...                                     1.0
5  democrat  ...                                     1.0
6  democrat  ...                                     1.0

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [110]:
dems_sample = dem.dropna()
reps_sample = rep.dropna()

In [111]:
print(dems_sample.isnull().sum())
reps_sample.isnull().sum()

Party                                     0
handicapped-infants                       0
water-project-cost-sharing                0
adoption-of-the-budget-resolution         0
physician-fee-freeze                      0
el-salvador-aid                           0
religious-groups-in-schools               0
anti-satellite-test-ban                   0
aid-to-nicaraguan-contras                 0
mx-missile                                0
immigration                               0
synfuels-corporation-cutback              0
education-spending                        0
superfund-right-to-sue                    0
crime                                     0
duty-free-exports                         0
export-administration-act-south-africa    0
dtype: int64


Party                                     0
handicapped-infants                       0
water-project-cost-sharing                0
adoption-of-the-budget-resolution         0
physician-fee-freeze                      0
el-salvador-aid                           0
religious-groups-in-schools               0
anti-satellite-test-ban                   0
aid-to-nicaraguan-contras                 0
mx-missile                                0
immigration                               0
synfuels-corporation-cutback              0
education-spending                        0
superfund-right-to-sue                    0
crime                                     0
duty-free-exports                         0
export-administration-act-south-africa    0
dtype: int64

In [116]:
dems_sample.mean(axis=0)


handicapped-infants                       0.588710
water-project-cost-sharing                0.451613
adoption-of-the-budget-resolution         0.854839
physician-fee-freeze                      0.048387
el-salvador-aid                           0.201613
religious-groups-in-schools               0.443548
anti-satellite-test-ban                   0.766129
aid-to-nicaraguan-contras                 0.830645
mx-missile                                0.790323
immigration                               0.532258
synfuels-corporation-cutback              0.508065
education-spending                        0.129032
superfund-right-to-sue                    0.290323
crime                                     0.346774
duty-free-exports                         0.596774
export-administration-act-south-africa    0.943548
dtype: float64

In [118]:
reps_sample.mean(axis=0)

handicapped-infants                       0.212963
water-project-cost-sharing                0.472222
adoption-of-the-budget-resolution         0.157407
physician-fee-freeze                      0.990741
el-salvador-aid                           0.953704
religious-groups-in-schools               0.870370
anti-satellite-test-ban                   0.268519
aid-to-nicaraguan-contras                 0.148148
mx-missile                                0.138889
immigration                               0.574074
synfuels-corporation-cutback              0.157407
education-spending                        0.851852
superfund-right-to-sue                    0.842593
crime                                     0.981481
duty-free-exports                         0.111111
export-administration-act-south-africa    0.666667
dtype: float64

In [121]:
##this is an issue where dems support the bill more
dems_sample['adoption-of-the-budget-resolution'].mean()

0.8548387096774194

In [122]:
reps_sample['adoption-of-the-budget-resolution'].mean()

0.1574074074074074

1)null hypothesis:

dems support this bill at the same level that republics do

$H_0: \overline{x}_r == \overline{x}_d$



2)alternative hypothesis:
dems support this bill in a different proportion than reps

$H_a: \overline{x}_r \neq \overline{x}_d$


3)confidence level 99%

In [125]:
stats.ttest_ind(dems_sample['adoption-of-the-budget-resolution'],reps_sample['adoption-of-the-budget-resolution'])

Ttest_indResult(statistic=14.742943396634677, pvalue=4.458031105913572e-35)

4)conclusion:

due to a t-statistic of 14.74 and p-value of ~0 we reject the null hypothesis that democrats and republics support the adoption of the bughet resolution bill at the same rate, and suggest the alternative that the dems supports the bill at a higher rate.

In [119]:
##this is an issue where reps support the bill more
dems_sample['crime'].mean()

0.3467741935483871

In [120]:
reps_sample['crime'].mean()

0.9814814814814815

1)null hypothesis:

dems support this bill at the same level that republics do

$H_0: \overline{x}_r == \overline{x}_d$



2)alternative hypothesis:

reps support this bill in a different proportion than dems

$H_a: \overline{x}_r \neq \overline{x}_d$


3)confidence level 99%

In [123]:
stats.ttest_ind(dems_sample['crime'],reps_sample['crime'])

Ttest_indResult(statistic=-13.34079793707258, pvalue=1.8791646941359472e-30)

4)conclusion:

due to a t-statistic of 16.34 and p-value of ~0 we reject the null hypothesis that democrats and republics support the crime bill at the same rate, and suggest the alternative that the reps supports the bill at a higher rate.


In [126]:
##this is an issue where reps and dems suppor the same
dems_sample['water-project-cost-sharing'].mean()

0.45161290322580644

In [127]:
reps_sample['water-project-cost-sharing'].mean()

0.4722222222222222

1)null hypothesis:

dems support this bill at the same level that republics do

$H_0: \overline{x}_r == \overline{x}_d$



2)alternative hypothesis:

reps support this bill in a different proportion than dems

$H_a: \overline{x}_r \neq \overline{x}_d$


3)confidence level 90%

In [129]:
stats.ttest_ind(reps_sample['water-project-cost-sharing'],dems_sample['water-project-cost-sharing'])

Ttest_indResult(statistic=0.3128208943045492, pvalue=0.7547001979801332)

4)conclusion:

due to a t-statistic of 0.312 and p-value of 0.754 we fail to reject the null hypothesis that democrats and republics support the crime bill at the same rate.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
### YOUR WORK HERE



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE